flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java

```diff
@@ -218,11 +205,61 @@ public Collection<FreeSlotInfo> getFreeSlotsInformation() {
         return freeSlotInfos;
     }
 
+    public double getTaskExecutorUtilization(ResourceID resourceId) {
+        Set<AllocationID> slots = slotsPerTaskExecutor.get(resourceId);
+        Preconditions.checkNotNull(slots, "There is no slots on %s", resourceId);
+
+        return (double) (slots.size() - freeSlots.getSlotsOfTaskExecutor(resourceId).size())
+                / slots.size();
+    }
+
     @Override
     public Collection<? extends SlotInfo> getAllSlotsInformation() {
         return registeredSlots.values();
     }
 
+    private static final class FreeSlots {
+        /** Map containing all free slots and since when they are free. */
+        private final Map<AllocationID, Long> freeSlotsSince = new HashMap<>();
+
+        /** Index containing a mapping between TaskExecutors and their free slots. */
+        private final Map<ResourceID, Set<AllocationID>> freeSlotsPerTaskExecutor = new HashMap<>();
```

Comment:
It seems that we only need the size of `Set<AllocationID>` here, how about changing it to `Map<ResourceID, Long>` ?